### ISSUE TITLE
length should be a property, not a function call

### ISSUE THREAD

**Issue by StoneCypher** [StoneCypher/circular_buffer_js#35]
this implies that full, empty, capacity, and available should be, too

thinking maybe i should rename empty to isEmpty, so it's not interpreted as a verb, too.  would suggest I should do the same for full

@drohen - would it be inconvenient for you if i changed the shape of that api?

**Comment by louisfoster** [comment]
Yeah was looking at the length method yesterday and wondering if this actually makes sense, but I'm wondering its utility if it doesn't implement any underlying iterator/collection class if I'm not mistaken (ie you've created your own data structure, therefore you can use whatever API is suitable for its use, not one that needs to conform to expectations to use in something like a for loop).

That being said, is there a reference API you're basing the move to verbs/properties for the other methods other than length? Go with your gut on this one I guess, as long as it makes reasonable sense and is documented, that's all that matters.
I'm a method kind of dev, but I'm used to the JS "length" property as the exception.

Other than that just keep in mind use cases of when something will be referenced, ie if I call something in a loop, will that value change between iterations if I call it at the beginning of a 
loop, potentially causing loops to go on forever?

I appreciate you asking. Just make sure if you do change the API to maybe throw errors when methods aren't available and do a major version bump in npm.

**Comment by louisfoster** [comment]
Ah I didn't see the name of the linked issue, which kind of answers my first question. There's 2 kinds of iterator/collection, ones like arrays that are kind of like objects that use numbers as keys, and iterators that use functions like next() and return values like done: true. So I'm wondering then what API you're trying to get closest to.

**Comment by drohen** [comment]
I posted earlier with my alter ego account, so I'm reposting this for consistency of the conversation, sorry about any confusion:

Yeah was looking at the length method yesterday and wondering if this actually makes sense, but I'm wondering its utility if it doesn't implement any underlying iterator/collection class if I'm not mistaken (ie you've created your own data structure, therefore you can use whatever API is suitable for its use, not one that needs to conform to expectations to use in something like a for loop).

That being said, is there a reference API you're basing the move to verbs/properties for the other methods other than length? Go with your gut on this one I guess, as long as it makes reasonable sense and is documented, that's all that matters.
I'm a method kind of dev, but I'm used to the JS "length" property as the exception.

Other than that just keep in mind use cases of when something will be referenced, ie if I call something in a loop, will that value change between iterations if I call it at the beginning of a
loop, potentially causing loops to go on forever?

I appreciate you asking. Just make sure if you do change the API to maybe throw errors when methods aren't available and do a major version bump in npm.


Ah I didn't see the name of the linked issue, which kind of answers my first question. There's 2 kinds of iterator/collection, ones like arrays that are kind of like objects that use numbers as keys, and iterators that use functions like next() and return values like done: true. So I'm wondering then what API you're trying to get closest to.

**Comment by StoneCypher** [comment]
i'm just trying to support principle of least surprise

i keep calling `.length` instead of `.length()` and it's making me feel dumb, is all

**Comment by StoneCypher** [comment]
> ie if I call something in a loop, will that value change between iterations if I call it at the beginning of a loop, potentially causing loops to go on forever?

length will change when the contents of the container change

if you change the contents of the container during the loop, length will also change

length is a non-mutating call and will not cause its own changes

**Comment by drohen** [comment]
> i'm just trying to support principle of least surprise
Nothing surprises me anymore

Any case, I'd only recommend changing the length method to a property but leave the others. But if you do change them, I don't mind, my usage of your lib isn't deployed to production yet. Good news its that it seems to working nicely though. Although I caught myself out accidentally messing up capacity/available which led to infinite recursion. Not sure if thats a design thing or user error.

**Comment by StoneCypher** [comment]
> Although I caught myself out accidentally messing up capacity/available which led to infinite recursion. Not sure if thats a design thing or user error.

I would like to understand more about how this happened please

**Comment by StoneCypher** [comment]
In general:

```typescript
const foo = new circular_buffer<number>(5);
foo.push(1);
foo.push(2);

console.log(`
  Capacity expects 5: ${foo.capacity()}, 
  available expects 3: ${foo.available()}, 
  size expects 2: ${foo.size()}
`);
```

`capacity/0` should always be the number from the constructor.  This will only change after `resize/1` is implemented.

`size/0` should be the count of values currently in the container.

`available/0` should be the number of times you can push before running out of space.

**Comment by drohen** [comment]
Actually I checked my code, I mixed up available and length. Probably a copy paste issue tbh. But I have 2 loops, one checking if there's "available" space (very useful), and another checking how many items I have in the buffer (ie length). These loops are practically identical, even storing the value in the same variable name, so I think I did actually originally copy and paste it and forgot to change it.
